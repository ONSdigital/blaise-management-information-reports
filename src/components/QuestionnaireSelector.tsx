import { ONSPanel, StyledForm } from "blaise-design-system-react-components";
import React, { ReactElement } from "react";

interface QuestionnaireSelectorProps {
    questionnaires: string[];
    selectedQuestionnaires: string[];
    setSelectedQuestionnaires: (string: string[]) => void;
    onSubmit: () => void;
}

function QuestionnaireSelector({
    questionnaires,
    selectedQuestionnaires,
    setSelectedQuestionnaires,
    onSubmit,
}: QuestionnaireSelectorProps): ReactElement {
    function handleSubmit(values: any) {
        setSelectedQuestionnaires(values.questionnaires);
        onSubmit();
    }

    function displayCheckboxes(items: string[]) {
        if (items.length === 0) {
            return <ONSPanel>No questionnaires found for given parameters.</ONSPanel>;
        }

        const fields = [
            {
                name: "questionnaires",
                type: "checkbox",
                initial_value: selectedQuestionnaires,
                validate: (values: string[]) => (values.length > 0 ? undefined : "At least one questionnaire must be selected"),
                checkboxOptions: items.map((name) => ({
                    id: name,
                    value: name,
                    label: name,
                    testid: name,
                })),
            },
        ];

        return <StyledForm fields={fields} submitLabel="Run report" onSubmitFunction={handleSubmit} />;
    }

    return (
        <div className="input-items">
            <div className="checkboxes__items">
                { displayCheckboxes(questionnaires) }
            </div>
        </div>
    );
}

export default QuestionnaireSelector;
