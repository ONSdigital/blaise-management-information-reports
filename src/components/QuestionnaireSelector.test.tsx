/**
 * @jest-environment jsdom
 */

import "@testing-library/jest-dom";
import QuestionnaireSelector from "./QuestionnaireSelector";
import React from "react";
import { cleanup, render, RenderResult } from "@testing-library/react";
import { fireEvent, screen } from "@testing-library/dom";
import { act } from "react-dom/test-utils";
import MockAdapter from "axios-mock-adapter";
import axios from "axios";

const mockAdapter = new MockAdapter(axios);

const questionnairesReturned = ["LMS2101_AA1", "LMS2101_BB1", "LMS2101_CC1"];
describe("QuestionnaireSelector tests", () => {
    let setQuestionnaires: (questionnaires: string[]) => void;
    let submit: () => void;
    let wrapper: RenderResult;
    
    beforeEach(async () =>{
        const axiosMock = new MockAdapter(axios);
        axiosMock.onPost("/api/questionnaires").reply(200, questionnairesReturned);
        setQuestionnaires = jest.fn();
        submit = jest.fn();
        await act(async () => {
            wrapper = renderComponent();
        });
    });

    afterEach(() => {
        mockAdapter.reset();
    });
    
    function renderComponent(){
        return render(
            <QuestionnaireSelector interviewer="Cal"
                startDate={ new Date("2022-05-04") }
                endDate={ new Date("2022-05-05") }
                surveyTla="LMS"
                questionnaires={ questionnairesReturned } 
                setQuestionnaires={ setQuestionnaires }
                submitFunction={ submit }/>
        );
    }

    it("matches snapshot", async () => {
        mockAdapter.onPost("/api/questionnaires").reply(200, questionnairesReturned);
        await screen.findByText("LMS2101_AA1");
        expect(wrapper).toMatchSnapshot();
    });

    describe("it renders correctly", () =>{
        it("displays a 'Select All' button", () => {
            expectSelectAllToBeDefined();
        });
        it("displays all available questionnaires", () =>{
            expectAvailableQuestionnairesToBeDefined(wrapper, questionnairesReturned);
        });
        it("displays all available questionnaires", () =>{
            expectCheckboxesForEachQuestionnaireToBeDefined(wrapper, questionnairesReturned);
        });
    });

    describe("when a questionnaire is selected", () => {
        it("displays a tick in the checkbox", () =>{
            expectTickWhenAQuestionnaireIsSelected(wrapper);
        });
    });

    describe("when 'Select all' is selected", () => {
        it("selects all questionnaires", () =>{
            expectSelectAllToSelectAllQuestionnaires(wrapper, questionnairesReturned);
        });
        it("changes to unselect all", () =>{
            expectSelectAllToChangeToUnselectAll(wrapper);
        });
    });

    describe("when 'Unselect all' is selected", () => {
        it("unselects all questionnaires", () =>{
            expectUnselectAllToUnselectAllQuestionnaires(wrapper, questionnairesReturned);
        });
        it("changes to select all", () =>{
            expectUnselectAllToChangeSelectAll(wrapper);
        });
    });

    afterAll(() => {
        cleanup();
    });
});

function expectSelectAllToBeDefined(){
    expect(screen.getByRole("button", { name: "Unselect All following checkboxes" })).toBeVisible();
}

function expectAvailableQuestionnairesToBeDefined(wrapper: RenderResult, questionnairesReturned: string[]){
    questionnairesReturned.forEach((questionnaire) => {
        expect(wrapper.getByText(questionnaire)).toBeVisible();
    });
}

function expectCheckboxesForEachQuestionnaireToBeDefined(wrapper: RenderResult, questionnairesReturned: string[]){
    questionnairesReturned.forEach((questionnaire) => {
        expect(wrapper.getByRole("checkbox", { name: questionnaire })).toBeVisible();
    });
}
function expectTickWhenAQuestionnaireIsSelected(wrapper: RenderResult){
    const checkBox = wrapper.getByRole("checkbox", { name: "LMS2101_AA1" }, { checked: true });
    fireEvent.click(checkBox);
    expect(wrapper.getByRole("checkbox", { name: "LMS2101_AA1" }, { checked: false }));
}

async function expectSelectAllToSelectAllQuestionnaires(wrapper: RenderResult, questionnairesReturned: string[]){
    await act(async () => {
        fireEvent.click(await screen.findByText(/Select all/));
    });
    questionnairesReturned.forEach((questionnaire) => {
        expect(wrapper.getByRole("checkbox", { name: questionnaire }, { checked: true }));
    });
}

async function expectSelectAllToChangeToUnselectAll(wrapper: RenderResult) {
    await act(async () => {
        fireEvent.click(await screen.findByText(/Select all/));
    });
    expect(wrapper.getByText("Unselect all")).toBeVisible();
}

async function expectUnselectAllToUnselectAllQuestionnaires(wrapper: RenderResult, questionnairesReturned: string[]) {
    await act(async () => {
        fireEvent.click(await screen.findByText(/Select all/));
        fireEvent.click(await screen.findByText(/Unselect all/));
    });
    questionnairesReturned.forEach((questionnaire) => {
        expect(wrapper.getByRole("checkbox", { name: questionnaire }, { checked: false }));
    });
}

async function expectUnselectAllToChangeSelectAll(wrapper: RenderResult) {
    await act(async () => {
        fireEvent.click(await screen.findByText(/Select all/));
        fireEvent.click(await screen.findByText(/Unselect all/));
    });
    expect(wrapper.getByText("Select all")).toBeVisible();
}
