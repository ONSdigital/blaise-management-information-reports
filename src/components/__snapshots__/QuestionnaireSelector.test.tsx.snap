// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionnaireSelector tests matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="input-items"
      >
        <div
          class="checkboxes__items"
        >
          <form
            action="#"
          >
            <fieldset
              class="fieldset"
            >
              <legend
                class="fieldset__legend"
              />
              <button
                class="btn u-mb-s js-auto-selector btn--small btn--secondary"
                type="button"
              >
                <span
                  class="btn__inner"
                >
                  <span
                    class="js-button-text"
                  >
                    Select All
                  </span>
                  <span
                    class="u-vh"
                  >
                     following checkboxes
                  </span>
                </span>
              </button>
              <div
                class="checkboxes__items"
                id="questionnaires"
              >
                <p
                  class="checkboxes__item"
                >
                  <span
                    class="checkbox"
                  >
                    <input
                      checked=""
                      class="checkbox__input js-checkbox"
                      id="LMS2101_AA1"
                      initial_value="LMS2101_AA1"
                      name="questionnaires"
                      type="checkbox"
                      value="LMS2101_AA1"
                    />
                    <label
                      class="checkbox__label "
                      for="LMS2101_AA1"
                      id="LMS2101_AA1-label"
                    >
                      LMS2101_AA1
                    </label>
                  </span>
                </p>
                <br />
                <p
                  class="checkboxes__item"
                >
                  <span
                    class="checkbox"
                  >
                    <input
                      class="checkbox__input js-checkbox"
                      id="LMS2101_BB1"
                      initial_value="LMS2101_AA1"
                      name="questionnaires"
                      type="checkbox"
                      value="LMS2101_BB1"
                    />
                    <label
                      class="checkbox__label "
                      for="LMS2101_BB1"
                      id="LMS2101_BB1-label"
                    >
                      LMS2101_BB1
                    </label>
                  </span>
                </p>
                <br />
                <p
                  class="checkboxes__item"
                >
                  <span
                    class="checkbox"
                  >
                    <input
                      class="checkbox__input js-checkbox"
                      id="LMS2101_CC1"
                      initial_value="LMS2101_AA1"
                      name="questionnaires"
                      type="checkbox"
                      value="LMS2101_CC1"
                    />
                    <label
                      class="checkbox__label "
                      for="LMS2101_CC1"
                      id="LMS2101_CC1-label"
                    >
                      LMS2101_CC1
                    </label>
                  </span>
                </p>
                <br />
              </div>
            </fieldset>
            <br />
            <button
              class="btn "
              data-testid="submit-button"
              type="submit"
            >
              <span
                class="btn__inner"
              >
                Run report
              </span>
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="input-items"
    >
      <div
        class="checkboxes__items"
      >
        <form
          action="#"
        >
          <fieldset
            class="fieldset"
          >
            <legend
              class="fieldset__legend"
            />
            <button
              class="btn u-mb-s js-auto-selector btn--small btn--secondary"
              type="button"
            >
              <span
                class="btn__inner"
              >
                <span
                  class="js-button-text"
                >
                  Select All
                </span>
                <span
                  class="u-vh"
                >
                   following checkboxes
                </span>
              </span>
            </button>
            <div
              class="checkboxes__items"
              id="questionnaires"
            >
              <p
                class="checkboxes__item"
              >
                <span
                  class="checkbox"
                >
                  <input
                    checked=""
                    class="checkbox__input js-checkbox"
                    id="LMS2101_AA1"
                    initial_value="LMS2101_AA1"
                    name="questionnaires"
                    type="checkbox"
                    value="LMS2101_AA1"
                  />
                  <label
                    class="checkbox__label "
                    for="LMS2101_AA1"
                    id="LMS2101_AA1-label"
                  >
                    LMS2101_AA1
                  </label>
                </span>
              </p>
              <br />
              <p
                class="checkboxes__item"
              >
                <span
                  class="checkbox"
                >
                  <input
                    class="checkbox__input js-checkbox"
                    id="LMS2101_BB1"
                    initial_value="LMS2101_AA1"
                    name="questionnaires"
                    type="checkbox"
                    value="LMS2101_BB1"
                  />
                  <label
                    class="checkbox__label "
                    for="LMS2101_BB1"
                    id="LMS2101_BB1-label"
                  >
                    LMS2101_BB1
                  </label>
                </span>
              </p>
              <br />
              <p
                class="checkboxes__item"
              >
                <span
                  class="checkbox"
                >
                  <input
                    class="checkbox__input js-checkbox"
                    id="LMS2101_CC1"
                    initial_value="LMS2101_AA1"
                    name="questionnaires"
                    type="checkbox"
                    value="LMS2101_CC1"
                  />
                  <label
                    class="checkbox__label "
                    for="LMS2101_CC1"
                    id="LMS2101_CC1-label"
                  >
                    LMS2101_CC1
                  </label>
                </span>
              </p>
              <br />
            </div>
          </fieldset>
          <br />
          <button
            class="btn "
            data-testid="submit-button"
            type="submit"
          >
            <span
              class="btn__inner"
            >
              Run report
            </span>
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
